src/module/rules/rule-element/temp-hp.ts

```diff
@@ -1,29 +1,56 @@
 import { ActorType } from "@actor/data/index.ts";
 import { ChatMessagePF2e } from "@module/chat-message/index.ts";
 import { isObject } from "@util";
-import { RuleElementOptions } from "./base.ts";
-import { RuleElementData, RuleElementPF2e, RuleElementSource } from "./index.ts";
+import { RuleElementOptions, RuleElementPF2e, RuleElementSchema } from "./index.ts";
+import type { BooleanField, SchemaField } from "types/foundry/common/data/fields.d.ts";
+import { ResolvableValueField, RuleElementSource } from "./data.ts";
 
 /**
  * @category RuleElement
  */
-class TempHPRuleElement extends RuleElementPF2e {
+class TempHPRuleElement extends RuleElementPF2e<TempHPRuleSchema> {
     static override validActorTypes: ActorType[] = ["character", "npc", "familiar"];
 
-    constructor(data: TempHPSource, options: RuleElementOptions) {
+    static override defineSchema(): TempHPRuleSchema {
+        const { fields } = foundry.data;
+        return {
+            ...super.defineSchema(),
+            value: new ResolvableValueField({ required: true, nullable: false }),
+            events: new fields.SchemaField(
+                {
+                    onCreate: new fields.BooleanField({ required: false, nullable: false, initial: undefined }),
+                    onTurnStart: new fields.BooleanField({ required: false, nullable: false, initial: undefined }),
+                },
+                { required: false, nullable: false, initial: undefined }
+            ),
+        };
+    }
+
+    constructor(data: RuleElementSource, options: RuleElementOptions) {
+        // SchemaField coerces invalid source values into an object so these checks must happen before the super call
+        if ("events" in data) {
+            if (!isObject<SourceFromSchema<TempHPEventsSchema>>(data.events)) {
+                throw Error("events must be an object");
+            } else if (data.events.onCreate === undefined && data.events.onTurnStart === undefined) {
+                throw Error("events object must have an onCreate or onTurnStart value or be omitted");
+            }
+        }
         super(data, options);
 
-        /** Whether the temporary hit points are immediately applied */
-        this.data.onCreate = !!(this.data.onCreate ?? true);
-        /** Whether the temporary hit points renew each round */
-        this.data.onTurnStart = !!this.data.onTurnStart;
+        // Set default values if events was omitted
+        if (!this.events) {
+            this.events = {
+                onCreate: true,
+                onTurnStart: false,
+            };
+        }
```

Comment:
Handle all this from the schema. If it doesn't validate until the migration, then it'll fail, and that's fine.

Reply:
I had some problems getting the schema field to behave like I wanted. I had it set up with default values at one point but setting the `events` value to a string will not trigger a validation error because the field just converts that to the default object.